File: web/src/protocol.ts
export const PROTOCOL_VERSION = 1;
export const OBJECT_ID = 'demo-cube';
export const MAX_MESSAGE_BYTES = 64 * 1024;

export type MessageKind = 'ready' | 'snapshot' | 'patch' | 'ack' | 'ping' | 'pong' | 'error';
export type Origin = 'host' | 'browser';

export interface LamportStamp {
  counter: number;
  origin: Origin;
}

export interface PropertyChange {
  key: string;
  value: unknown;
  stamp: LamportStamp;
}

export interface PatchPayload {
  changes: PropertyChange[];
}

export interface SyncEnvelope {
  protocolVersion: number;
  kind: MessageKind;
  objectId: string;
  origin: Origin;
  sequence: number;
  correlationId?: string;
  payload: unknown;
}

export function parseEnvelope(json: string): SyncEnvelope {
  if (new TextEncoder().encode(json).length > MAX_MESSAGE_BYTES) throw new Error('Message exceeds 64 KiB.');
  const value = JSON.parse(json) as Partial<SyncEnvelope>;
  const kinds: MessageKind[] = ['ready', 'snapshot', 'patch', 'ack', 'ping', 'pong', 'error'];
  if (value.protocolVersion !== PROTOCOL_VERSION) throw new Error('Unsupported protocol version.');
  if (!value.kind || !kinds.includes(value.kind)) throw new Error('Unknown message kind.');
  if (value.objectId !== OBJECT_ID) throw new Error('Unexpected object id.');
  if (value.origin !== 'host' && value.origin !== 'browser') throw new Error('Unexpected origin.');
  if (!Number.isSafeInteger(value.sequence) || value.sequence! < 0) throw new Error('Invalid sequence.');
  return value as SyncEnvelope;
}

export function compareStamp(left: LamportStamp, right?: LamportStamp): number {
  if (!right) return 1;
  if (left.counter !== right.counter) return left.counter - right.counter;
  return left.origin.localeCompare(right.origin);
}

export function envelope(kind: MessageKind, sequence: number, payload: unknown, correlationId?: string): SyncEnvelope {
  return { protocolVersion: PROTOCOL_VERSION, kind, objectId: OBJECT_ID, origin: 'browser', sequence, correlationId, payload };
}

File: src/ThreeJsSync.Core/Protocol.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Serialization;

namespace ThreeJsSync.Core
{
    public static class Protocol
    {
        public const int Version = 1;
        public const int MaxMessageBytes = 64 * 1024;
        public const string DefaultObjectId = "demo-cube";

        public static readonly JsonSerializerSettings JsonSettings = new JsonSerializerSettings
        {
            ContractResolver = new CamelCasePropertyNamesContractResolver(),
            NullValueHandling = NullValueHandling.Ignore,
            Formatting = Formatting.None
        };

        public static string Serialize(SyncEnvelope envelope) => JsonConvert.SerializeObject(envelope, JsonSettings);

        public static SyncEnvelope Deserialize(string json)
        {
            if (json == null) throw new ProtocolException("Message is null.");
            if (System.Text.Encoding.UTF8.GetByteCount(json) > MaxMessageBytes)
                throw new ProtocolException("Message exceeds 64 KiB.");
            try
            {
                var envelope = JsonConvert.DeserializeObject<SyncEnvelope>(json, JsonSettings);
                if (envelope == null) throw new ProtocolException("Message is empty.");
                ValidateEnvelope(envelope);
                return envelope;
            }
            catch (JsonException ex)
            {
                throw new ProtocolException("Malformed JSON message.", ex);
            }
        }

        public static void ValidateEnvelope(SyncEnvelope envelope)
        {
            if (envelope.ProtocolVersion != Version) throw new ProtocolException("Unsupported protocol version.");
            if (string.IsNullOrWhiteSpace(envelope.Kind) || !MessageKinds.All.Contains(envelope.Kind))
                throw new ProtocolException("Unknown message kind.");
            if (envelope.ObjectId != DefaultObjectId) throw new ProtocolException("Unexpected object id.");
            if (envelope.Origin != "host" && envelope.Origin != "browser")
                throw new ProtocolException("Unexpected origin.");
            if (envelope.Sequence < 0) throw new ProtocolException("Sequence cannot be negative.");
        }
    }

    public static class MessageKinds
    {
        public const string Ready = "ready";
        public const string Snapshot = "snapshot";
        public const string Patch = "patch";
        public const string Ack = "ack";
        public const string Ping = "ping";
        public const string Pong = "pong";
        public const string Error = "error";
        public static readonly HashSet<string> All = new HashSet<string>
        {
            Ready, Snapshot, Patch, Ack, Ping, Pong, Error
        };
    }

    public sealed class SyncEnvelope
    {
        public int ProtocolVersion { get; set; } = Protocol.Version;
        public string Kind { get; set; }
        public string ObjectId { get; set; } = Protocol.DefaultObjectId;
        public string Origin { get; set; }
        public long Sequence { get; set; }
        public string CorrelationId { get; set; }
        public JToken Payload { get; set; }
    }

    public sealed class PatchPayload
    {
        public IList<PropertyChange> Changes { get; set; } = new List<PropertyChange>();
    }

    public sealed class PropertyChange
    {
        public string Key { get; set; }
        public JToken Value { get; set; }
        public LamportStamp Stamp { get; set; }
    }

    public sealed class LamportStamp : IComparable<LamportStamp>
    {
        public long Counter { get; set; }
        public string Origin { get; set; }

        public int CompareTo(LamportStamp other)
        {
            if (other == null) return 1;
            var counter = Counter.CompareTo(other.Counter);
            return counter != 0 ? counter : string.CompareOrdinal(Origin, other.Origin);
        }

        public LamportStamp Copy() => new LamportStamp { Counter = Counter, Origin = Origin };
    }

    public sealed class AckPayload
    {
        public long Sequence { get; set; }
        public string Checksum { get; set; }
    }

    public sealed class ProtocolException : Exception
    {
        public ProtocolException(string message) : base(message) { }
        public ProtocolException(string message, Exception inner) : base(message, inner) { }
    }
}

